Add benchmarks mandelbrot and fannkuch

nategraf opened the pull request:
Adding mandelbrot and fankuch-redux from Benchmarks Game to
complete the representation of benchmarks from this list
[Benchmarks Game C# Benchmarks](http://benchmarksgame.alioth.debian.org/u64q/csharp.html)

@dotnet/rap-team

dnfclas commented:
@__nategraf__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all .NET Foundation-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

dnfclas commented:
__@nategraf__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

AndyAyersMS commented:
Can you add comments indicating which variants of the test you based these on?

Tests need to return an exit status of 100 or the test system will think they have failed. Ideally you'd set the status to 100 only after validating that the benchmark did the computation requested correctly. I'd encourage you to investigate and see if this is possible.

Also there are other benchmark config files in peer directories to the one you are changing. If we are moving benchmarks to netstandard 1.5 then I'd recommend moving the others at the same time, perhaps as a prerequisite change. Any reason we can't get all these to netcoreapp2.0?

nategraf commented:
- I added comments to clarify the exact source from which the new benchmarks are derived.
- I modified the Main function to return 100 instead of 0
- I created checksums which are checked after each iteration and will fail the benchmark if they do not match. The checksums were calculated using the source on Benchmarks Game from which the new benchmark was derived, and assumed to be correct based on the fact that Benchmark Games accepted it.

The bump to .NET standard 1.5 was due to the use of Assembly.GetEntryAssembly. Based on the current configuration of the tests this needed to be done for all benchmarks at once. (unless I just don't know the better way) It may be possible to bump everything to netcoreapp2.0, but 1.5 was chosen because it was the smallest step necessary.

AndyAyersMS commented:
Benchmark code looks ok.

Your tests are not building cleanly yet -- there is apparently a new dependence on a commandline package. Would be nice not to have this dependence.

nategraf commented:
After some discussion with @AndyAyersMS the Main functions within these two benchmarks has been rewritten to remove XunitPerformanceHarness thereby eliminating the new dependencies the previous commit introduced. 
The benefits of this approach, as opposed to making all test Main functions use XunitPerformanceHarness, that decided this change were:
- Faster verification speed, allowing more taxing performance testing without slowing down the verification tests
- Consistency of assembly listings between runs
- Assembly listings contain only benchmark code; not xunit/xunit-perf code
- Tests are easier to debug
- A purer focus on verification. The Main function has no effect on the performance aspects of the benchmark, so the Main's only purpose is to verify the algorithm calculates the correct result.

jorive commented on tests/src/JIT/Performance/CodeQuality/BenchmarksGame/mandelbrot/mandelbrot.csharp.cs:
```diff
@@ -25,6 +25,8 @@ namespace BenchmarksGame
 {
     public class Mandelbrot
     {
+        private static bool s_testFailed = false;
```
This can now be removed.

nategraf replied:
Fix in commit d1da9ab822b22f4fe26b6dda7c03942a19de17f3

AndyAyersMS commented:
Overall looks pretty good. Failure in the latest run looks unrelated; the tests you added passed.

How long does each iteration take when you run these under xunit-perf? We've been aiming at about 1sec per iteration. You may need to use the InnerIterationCount to do more repetitions in each timed segment.

nategraf commented:
Here are some example runs on my workstation. Currently the tests are targeted at n=4000 for mandelbrot and n=10 for fannkuch-redux

[Perf-Off-mandelbrot.txt](https://github.com/dotnet/coreclr/files/1166101/Perf-Off-mandelbrot.txt)
[Perf-Off-fannkuch-redux.txt](https://github.com/dotnet/coreclr/files/1166102/Perf-Off-fannkuch-redux.txt)

Currently Mandelbrot is at 850 ms on my workstation, and the test can be adjusted arbitrarily to hit a particular load
Fannkuch takes a small integer input and grow exponentially. Currently the test uses n=10 which runs in 150 ms and the next step up (n=11) runs in 2000 ms. 

@AndyAyersMS Do you want to see those parameters adjusted?

AndyAyersMS commented:
150ms is a bit on the low side. You could do something like this to amplify it a bit:
```C#
[Benchmark(InnerIterationCount = 7)]
[InlineData(10, 73196)
public static void Test(int n, int checksum)
{
    foreach (var iteration in Benchmark.Iterations)
    {
        int[] pfannkuchen = null;
        using (iteration.StartMeasurement())
        {
            for (int i = 0; i < Benchmark.InnerIterationCount; ++i)
            {
                pfannkuchen = Bench(n);
            }
        }
        Assert.Equal(checksum, pfannkuchen[0]);
    }
}
```
